homeassistant/components/bluetooth/active_update_coordinator.py

```diff
@@ -0,0 +1,122 @@
+"""A Bluetooth passive coordinator that collects data from advertisements but can also poll."""
+from __future__ import annotations
+
+import asyncio
+from collections.abc import Callable, Coroutine
+import logging
+import time
+from typing import Any, TypeVar
+
+from homeassistant.core import HomeAssistant, callback
+
+from . import BluetoothChange, BluetoothScanningMode, BluetoothServiceInfoBleak
+from .passive_update_processor import (
+    PassiveBluetoothDataUpdate,
+    PassiveBluetoothProcessorCoordinator,
+)
+
+_T = TypeVar("_T")
+
+
+class ActiveBluetoothProcessorCoordinator(PassiveBluetoothProcessorCoordinator):
+    """
+    A coordinator that parses passive data from advertisements but can also poll.
+
+    Every time an advertisement is received, needs_poll_method is called to work
+    out if a poll is needed. This should return True if it is and False if it is
+    note.
+
+    def needs_poll_method(svc_info: BluetoothServiceInfoBleak, last_poll: float | None) -> bool:
+        return True
+
+    If there has been no poll since HA started, `last_poll` will be None. Otherwise it is
+    the number of seconds since one was last attempted.
+
+    If a poll is needed, the coordinator will call poll_method. This is a coroutine.
+    It should return the same type of data as your update_method. The expectation is that
+    data from advertisements and from polling are being parsed and fed into a shared
+    object that represents the current state of the device.
+
+    async def poll_method(svc_info: BluetoothServiceInfoBleak) -> YourDataType:
+        return YourDataType(....)
+
+    BluetoothServiceInfoBleak.device contains a BLEDevice. You should use this in
+    your poll function, as it is the most efficient way to get a BleakClient.
+    """
+
+    def __init__(
+        self,
+        hass: HomeAssistant,
+        logger: logging.Logger,
+        address: str,
+        mode: BluetoothScanningMode,
+        update_method: Callable[[BluetoothServiceInfoBleak], _T],
+        needs_poll_method: Callable[[BluetoothServiceInfoBleak, float | None], bool],
+        poll_method: Callable[
+            [BluetoothServiceInfoBleak],
+            Coroutine[Any, Any, PassiveBluetoothDataUpdate[_T]],
+        ],
+    ) -> None:
+        """Initialize the processor."""
+        super().__init__(hass, logger, address, mode, update_method)
+
+        self._needs_poll_method = needs_poll_method
+        self._poll_method = poll_method
+        self._last_poll: float | None = None
+        self._last_poll_successful = True
+        self._poll_limiter = asyncio.Semaphore(2)
```

Comment:
Will this work for all integrations or should we use a debouncer as the normal DataUpdateCoordinator that integrations can override?

Reply:
I need to look at what they do, i'll come back to this in a few hours.

Reply:
Tried to mirror what the main DataUpdateCoordinator does in https://github.com/home-assistant/core/pull/76342/commits/c0c571bc3610863051128fa900a975307aa092d9.